db/src/timelines.rs

```diff
@@ -0,0 +1,176 @@
+            let t = t?;
+            if !tx_ids.contains(&t.0) {
+                bail!(DbErrorKind::TimelinesNotOnTail);
+            }
+            t.1
+        },
+        None => bail!(DbErrorKind::TimelinesInvalidTransactionIds)
+    };
+
+    while let Some(t) = rows.next() {
+        let t = t?;
+        if !tx_ids.contains(&t.0) {
+            bail!(DbErrorKind::TimelinesNotOnTail);
+        }
+        if t.1 != timeline {
+            bail!(DbErrorKind::TimelinesMixed);
+        }
+    }
+
+    Ok(timeline)
+}
+
+fn move_transactions_to(conn: &rusqlite::Connection, tx_ids: &[Entid], new_timeline: Entid) -> Result<()> {
+    // Move specified transactions over to a specified timeline.
+    conn.execute(&format!(
+        "UPDATE timelined_transactions SET timeline = {} WHERE tx IN ({})",
+            new_timeline,
+            ::repeat_values(tx_ids.len(), 1)
+        ), &(tx_ids.iter().map(|x| x as &rusqlite::types::ToSql).collect::<Vec<_>>())
+    )?;
+    Ok(())
+}
+
+fn eradicate_transaction(conn: &rusqlite::Connection, tx: Entid) -> Result<()> {
+    conn.execute("DELETE FROM timelined_transactions WHERE tx = ?", &[&tx])?;
+    Ok(())
+}
+
+/// Get terms after tx_id, and reverse them both in order (DESC tx) and in meaning (swap add & retract).
+fn reversed_terms_after(conn: &rusqlite::Connection, tx_id: Entid) -> Result<Vec<TermWithoutTempIds>> {
+    let mut stmt = conn.prepare("SELECT e, a, v, value_type_tag, tx, added FROM timelined_transactions WHERE tx > ? AND timeline = ? ORDER BY tx DESC")?;
+    let mut rows = stmt.query_and_then(&[&tx_id, &MAIN_TIMELINE], |row| -> Result<TermWithoutTempIds> {
+        let op = match row.get_checked(5)? {
+            true => OpType::Retract,
+            false => OpType::Add
+        };
+        Ok(Term::AddOrRetract(
+            op,
+            KnownEntid(row.get_checked(0)?),
+            row.get_checked(1)?,
+            TypedValue::from_sql_value_pair(row.get_checked(2)?, row.get_checked(3)?)?,
+        ))
+    })?;
+
+    let mut terms = vec![];
+
+    while let Some(row) = rows.next() {
+        terms.push(row?);
+    }
+    Ok(terms)
+}
+
+pub fn move_from_main_timeline(conn: &rusqlite::Connection, schema: &Schema,
+    mut partition_map: PartitionMap, tx_ids: &[Entid], new_timeline: Entid
+    ) -> Result<(PartitionMap, Option<Schema>)> {
+
+    if new_timeline == MAIN_TIMELINE {
+        bail!(DbErrorKind::NotYetImplemented(format!("Can't move transactions to main timeline")));
+    }
+
+    let smallest_tx = match tx_ids.iter().min() {
+        Some(e) => *e,
+        None => bail!(DbErrorKind::TimelinesNoTransactionsSupplied)
+    };
+
+    let timeline = current_timeline(conn, smallest_tx, tx_ids)?;
+
+    // TODO stil need to return schema and partitionmap...
+    if timeline != MAIN_TIMELINE {
+        bail!(DbErrorKind::TimelinesNotOnMain);
+    }
+
+    let reversed_terms = reversed_terms_after(conn, smallest_tx - 1)?;
+    let lowest_e = match reversed_terms.last() {

```

Comment:
Fundamentally, it's not possible to determine the set of known partitions, or the current "next index" of a partition, by examining `datoms` or any fixed length set of `transactions`.  You could have:
- tx 1 could allocate 1000 entids
- tx 2 ... N only reference the first, say, 10 entids
- tx N + 1 retracts entid 1000
You'll never "discover" the real range of valid entids.

Thus, the transactor needs to be able to tell you exactly the state of the partition map at each transaction.  I've started working on this and will push something up when I can, but you can also start thinking about this, if you'd like.